magma/types/handshake.py

```diff
@@ -36,14 +36,101 @@ def _maybe_add_data(cls, fields, T, qualifier):
     fields["data"] = qualifier(T)
 
 
-class ReadyValidKind(ProductKind):
+def _enq_error(self, value, when=True):
+    raise Exception(f"{type(self)} does not support enq")
+
+
+def _no_enq_error(self, value, when=True):
+    raise Exception(f"{type(self)} does not support no_enq")
+
+
+def _deq_error(self, value, when=True):
+    raise Exception(f"{type(self)} does not support deq")
+
+
+def _no_deq_error(self, value, when=True):
+    raise Exception(f"{type(self)} does not support no_deq")
+
+
+class HandShakeKind(ProductKind):
+    # Undirected
+    _qualifers_ = {"output": lambda x: x, "input": lambda x: x,
+                   "data": lambda x: x}
+
+    def _make_fields(cls, T):
+        """
+        Each HandShake implementation should define a dict of _field_names_ for
+        the signals (e.g. "ready" and "valid").  For consistency, we define the
+        output/input names in terms of a Producer (i.e. ReadyValid outputs a
+        valid and inputs a ready)
+
+        Each implementation should define a dict of _qualifiers_ used to
+        qualify the ports for producer, consumer, etc...
+        """
+        fields = {
+            cls._field_names_["handshake_output"]:
+                cls._qualifers_["output"](Bit),
+            cls._field_names_["handshake_input"]:
+                cls._qualifers_["input"](Bit)
+        }
+        _maybe_add_data(cls, fields, T, cls._qualifers_["data"])
+        return fields
+
     def __getitem__(cls, T: Union[Type, MagmaProtocol]):
-        fields = {"valid": Bit, "ready": Bit}
-        _maybe_add_data(cls, fields, T, lambda x: x)
-        return type(f"{cls}[{T}]", (cls, ), fields)
+        assert cls._field_names_ is not None, "Undefined field name impl"
+        fields = cls._make_fields(T)
+
+        bases = (cls, )
+        if hasattr(cls, "Base_T"):
+            # Qualified variants include the unqualified type as a base class
+            # e.g. Producer(ReadyValid[Bits[8]]) will be a subtype of
+            #      ReadyValid[Bits[8]]
+            bases += (cls.Base_T[T], )
+        name = f"{cls}[{T}]"
+
+        # Populate namespace for proper cache behavior (based on
+        # hwtypes.adt_meta logic)
+        ns = {}
+        ns['__module__'] = cls.__module__
+        name = f'{cls}[{T}]'
+        ns['__qualname__'] = name
+
+        # _cache_handler implemented in hwtypes.adt_meta
+        return cls._cache_handler(True, fields, name, bases, ns)
+
+    def __new__(mcs, name, bases, namespace, fields=None, **kwargs):
+        result = super().__new__(mcs, name, bases, namespace, fields, **kwargs)
+
+        if not result._field_names_:
+            # Skip for base class (variants already exist and pointer is not
```

Comment:
have a bunch of other nits on comments. at least, can we end comments in periods that are full sentences? otherwise if they're short include them inline, e.g. on L112, that can be after the if or then inline, e.g. `return result  # these don't need variants`